packages/core/src/parser/source_map.ts

```diff
@@ -0,0 +1,172 @@
+/*
+*                      Copyright 2020 Salto Labs Ltd.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with
+* the License.  You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+import _ from 'lodash'
+import { ElemID } from '@salto-io/adapter-api'
+import wu, { WuIterable } from 'wu'
+import { SourceRange, isSourceRange } from './internal/types'
+
+interface SourceMapEntry {
+  children: Record<string, SourceMapEntry>
+  value: SourceRange[]
+}
+
+const getFromPath = (
+  data: SourceMapEntry,
+  path: string[],
+  createIfMissing = false
+): SourceMapEntry | undefined => {
+  const [key, ...restOfPath] = path
+  if (data.children[key] === undefined && createIfMissing) {
+    data.children[key] = { children: {}, value: [] }
+  }
+  if (_.isEmpty(restOfPath)) {
+    return data.children[key]
+  }
+  return data.children[key]
+    ? getFromPath(data.children[key], restOfPath, createIfMissing)
+    : undefined
+}
+
+const mergeEntries = (src: SourceMapEntry, target: SourceMapEntry): void => {
+  src.value.push(...target.value)
+  // eslint-disable-next-line no-restricted-syntax
+  _.entries(target.children).forEach(([key, value]) => {
+    if (key in src.children) {
+      mergeEntries(src.children[key] as SourceMapEntry, value)
+    } else {
+      src.children[key] = value
+    }
+  })
+}
+
+const mountToPath = (
+  data: SourceMapEntry,
+  path: string[],
+  value: SourceMapEntry
+): void => {
+  const target = getFromPath(data, path, true)
+  if (target) {
+    mergeEntries(target, value)
+  }
+}
+
+const setToPath = (
+  data: SourceMapEntry,
+  path: string[],
+  value: SourceRange[]
+): void => {
+  const target = getFromPath(data, path, true)
+  if (target) {
+    target.value = value
+  }
+}
+
+const iterEntry = (
+  entry: SourceMapEntry,
+  prefix: string[] = []
+): WuIterable<[string, SourceRange[]]> => {
+  const childEntries = wu.entries(entry.children)
+    .map(([key, child]) => iterEntry(child, [...prefix, key]))
+    .flatten(true)
+  return _.isEmpty(entry.value)
+    ? childEntries
+    : wu.chain([[prefix.join(ElemID.NAMESPACE_SEPARATOR), entry.value]], childEntries)
+}
+
+export class SourceMap implements Map<string, SourceRange[]> {
+  [Symbol.toStringTag] = 'SourceMap'
+  private data: SourceMapEntry = { children: {}, value: [] }
+
+  constructor(entries: Iterable<[string, SourceRange[]]> = []) {
+    wu(entries).forEach(([key, value]: [string, SourceRange[]]) => this.set(key, value))
```

Comment:
pretty sure the types can be inferred
```suggestion
    wu(entries).forEach(([key, value]) => this.set(key, value))
```